### PR TITLE
Fix #2479 - Check migration at server start

### LINKED ISSUE #2479: OpenMetadata server should throw meaningful error incase of non-upgrade starts
users who are coming from 0.7 release might not run the ./bootstrap/bootstrap-storage.sh migrate before installing the 0.8 binaries and they will run into issues in the code or schema throwing error which are not helpful
1. flyway has DATABASE_CHANGE_LOG
image
2. we can see which version the schema currently running on
3. if we have newer version of sql files under bootstrap/sql/mysql we should throw an error saying users need to upgrade before starting the server
4. We can provide the link to upgrade https://docs.open-metadata.org/install/upgrade-openmetadata

### PR DESCRIPTION
**Pull request by pmbrull**
### Describe your changes :

This PR fixes https://github.com/open-metadata/OpenMetadata/issues/2479.

The idea behind this is being able to help users during the upgrade of OpenMetadata by checking the flyway versions on MySQL before starting the server.

As we are controlling flyway with the TableInitializer, here we can reuse its `validate` command and check the exit status to either show the message that users need to migrate, or we can continue normally.

One question I would have here is that with this behavior we will also be giving this notification to users that are installing OpenMetadata for the first time, as `openmetadata.sh` does not run `migrate-all`. @harshach are we OK with this behavior? If so, should we update [this](https://docs.open-metadata.org/install/run-in-production#4.-start-openmetadata) doc to show how to run the migration as well before starting the service?

Thanks!!

#
### Type of change :

- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

**Comment by harshach** [comment]
@pmbrull this could work, was expecting this to be part of CatalogApplication.java file. Users could use some other startup mechanism instead of using ./bin/catalog.sh then this approach might not work. But should be ok.
For users who are just installing for the first time then they won't be having DATABASE_CHANGE_LOG table as thats created by flyway. You'll be running this tool before invoking the migration so that case should work

**Comment by pmbrull** [comment]
> @pmbrull this could work, was expecting this to be part of CatalogApplication.java file. Users could use some other startup mechanism instead of using ./bin/catalog.sh then this approach might not work. But should be ok. For users who are just installing for the first time then they won't be having DATABASE_CHANGE_LOG table as thats created by flyway. You'll be running this tool before invoking the migration so that case should work

Hi @harshach, makes sense. I came up with this as I was following the docs myself. I can add the check on `CatalogApplication.java` as well

**Comment by pmbrull** [comment]
Hi @harshach, added the validation in the Catalog class as well. Let me know if this is the approach you had in mind 🙏  Thanks!

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/EmbeddedMySqlSupport.java [hunk 1]
```diff
@@ -50,6 +50,7 @@ public void beforeAll(ExtensionContext extensionContext) {
               // TODO Remove hardcoding
               .dataSource(
                   "jdbc:mysql://localhost:3307/openmetadata_test_db?useSSL=false&serverTimezone=UTC", "test", "")
+              .table("DATABASE_CHANGE_LOG")
```
Comment (pmbrull) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/EmbeddedMySqlSupport.java:53]: Add the changelog table on the testing embedded server as well

File: bin/openmetadata.sh [hunk 2]
```diff
@@ -102,6 +103,28 @@ function getPID {
    return 0
 }
 
+# Check if the mysql migrations are up to date
+function validateDbMigrations {
```
Comment (harshach) [inline bin/openmetadata.sh:107]: @pmbrull do we need this function, give you have it in CatalogApplication?

Comment (pmbrull) [reply bin/openmetadata.sh]: Removed 🙏

**Review by harshach — APPROVED** [review]

**Pull request merged**